src/timedelta_isoformat/__init__.py

```diff
@@ -85,33 +98,47 @@ def _from_designators(duration: str) -> Components:
         in order of largest-to-smallest unit from left-to-right (with the exception of
         week measurements, which must be the only measurement in the string if present).
         """
-        date_tokens = iter(("Y", "years", "M", "months", "D", "days"))
-        time_tokens = iter(("H", "hours", "M", "minutes", "S", "seconds"))
-        week_tokens = iter(("W", "weeks"))
-
-        tokens, value = date_tokens, ""
+        date_context = ParsingContext(
+            {
+                "Y": "years",
+                "M": "months",
+                "D": "days",
+            }
+        )
+        time_context = ParsingContext(
+            {
+                "H": "hours",
+                "M": "minutes",
+                "S": "seconds",
+            }
+        )
+        week_context = ParsingContext(
+            {"W": "weeks"}
+        )
+
+        context, value = date_context, ""
         for char in duration:
             if char in _FORMAT:
                 value += char
                 continue
 
-            if char == "T" and tokens is not time_tokens:
-                tokens, value = time_tokens, ""
+            if char == "T" and context is not time_context:
+                context, value = time_context, ""
                 continue
 
-            if char == "W" and tokens is date_tokens:
-                tokens = week_tokens
+            if char == "W" and context is date_context:
+                context = week_context
                 pass
 
             # Note: this advances and may exhaust the token iterator
-            if char not in tokens:
+            if char not in context.remaining_tokens:
                 raise ValueError(f"unexpected character '{char}'")
 
-            yield value, next(tokens), None
+            yield value, context.units[char], None
             value = ""
 
-        weeks_parsed = next(week_tokens, None) != "W"
-        time_parsed = next(time_tokens, None) != "H" or next(date_tokens, None) != "Y"
+        weeks_parsed = next(week_context.remaining_tokens, None) != "W"
+        time_parsed = next(time_context.remaining_tokens, None) != "H" or next(date_context.remaining_tokens, None) != "Y"
```

Comment:
Reviewing your code has refreshed my understanding of this section of the module (it's been a while).

I'd currently translate what it does as:

"Check whether any of the parser contexts have advanced from their original state during input parsing"

That could help to figure out how -- and whether -- we could refactor it into more-readable and/or more-efficient alternative implementations.